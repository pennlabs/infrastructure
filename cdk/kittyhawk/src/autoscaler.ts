// Mostly borrowed from https://github.com/awslabs/cdk8s/blob/master/examples/typescript/podinfo/lib/autoscaler.ts

import { Construct } from 'constructs';
import { KubeHorizontalPodAutoscalerV2Beta2 as HorizontalPodAutoscalerApiObject, MetricSpec, Quantity } from '../imports/k8s';

export interface ScaleTarget {
  readonly apiVersion: string;
  readonly kind: string;
  readonly name: string;
}

export interface AutoscalingProps {

  /**
     * target CPU usage per pod
     */
  readonly cpu?: number;

  /**
     * target memory usage per pod
     */
  readonly memory?: number;

  /**
     * target requests per second per pod
     */
  readonly requests?: number;

  /**
     * @default 10
     */
  readonly maxReplicas?: number;

  /**
     * @default 2
     */
  readonly minReplicas?: number;
}

export interface AutoscalerProps extends AutoscalingProps {
  /**
     * The scaling target.
     */
  readonly target: ScaleTarget;
}

export class Autoscaler extends Construct {
  constructor(scope: Construct, appname: string, props: AutoscalerProps) {
    super(scope, `autoscaler-${appname}`);

    const metrics = new Array<MetricSpec>();

    if (props.cpu) {
      metrics.push({
        type: 'Resource',
        resource: {
          name: 'cpu',
          target: {
            type: 'Utilization',
            averageUtilization: props.cpu,
          },
        },
      });
    }

    if (props.memory) {
      metrics.push({
        type: 'Resource',
        resource: {
          name: 'memory',
          target: {
            type: 'AverageValue',
            averageValue: Quantity.fromNumber(props.memory),
          },
        },
      });
    }

    if (props.requests) {
      metrics.push({
        type: 'Pod',
        pods: {
          metric: {
            name: 'http_requests',
          },
          target: {
            type: 'AverageValue',
            averageValue: Quantity.fromNumber(props.requests),
          },
        },
      });
    }

    new HorizontalPodAutoscalerApiObject(this, 'default', {
      metadata: {
        name: appname,
      },
      spec: {
        scaleTargetRef: {
          apiVersion: props.target.apiVersion,
          kind: props.target.kind,
          name: props.target.name,
        },
        minReplicas: props.minReplicas || 2,
        maxReplicas: props.maxReplicas || 10,
        metrics,
      },
    });
  }
}