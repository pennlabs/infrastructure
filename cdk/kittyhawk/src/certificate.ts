import { Construct } from 'constructs';
import { Certificate as CertApiObject } from '../imports/cert-manager.io/certificate';
import { IngressProps } from './ingress';

export class Certificate extends Construct {
  constructor(scope: Construct, appname: string, props: IngressProps) {
    super(scope, `certificate-${appname}`);

    // Only generate certificates if an ingress is defined
    if (props.ingress) {
      // We want to generate a certificate for each host
      for (const h of props.ingress) {
        // Regex to compute the apex domain
        const apex_domain = h.host.match(/[\w-]+\.[\w]+$/g)
        if (apex_domain != null) {
          let host_string = apex_domain[0].split('.').join('-');
          new CertApiObject(this, `certificate-${appname}-${host_string}`, {
            metadata: {
              name: host_string,
            },
            spec: {
              secretName: host_string.concat('-tls'),
              dnsNames: [`${apex_domain[0]}`, `*.${apex_domain[0]}`],
              issuerRef: {
                name: 'wildcard-letsencrypt-prod',
                kind: 'ClusterIssuer',
                group: 'cert-manager.io',
              },
            },
          });
        } else
          throw `Certificate construction failed: apex domain regex failed on ${h}`
      }
    }
  }
}