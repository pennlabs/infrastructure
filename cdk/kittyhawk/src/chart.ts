import { Construct } from 'constructs';
import { App, Chart as ChartApiObject } from 'cdk8s';

export class Chart extends ChartApiObject {
  constructor(scope: Construct, name: string, chartBuilder: (scope: Construct) => void) {
    // grab the release name from env vars
    super(scope, name, {
      labels : {
        release: (process.env.RELEASE_NAME || 'undefined_release')}}); 
    chartBuilder(this);
  }
}

export function synth (chartBuilder: (scope: Construct) => void) {
  const app = new App();
  new Chart(app, 'kittyhawk', chartBuilder);
  app.synth();
}