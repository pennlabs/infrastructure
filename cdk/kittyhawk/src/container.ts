import { Volume as VolumeInterface, Container as ContainerInterface, ContainerPort, EnvFromSource, EnvVar, Probe as ProbeInterface, VolumeMount, SecretVolumeSource, HttpGetAction, ExecAction } from '../imports/k8s';


export interface ContainerProps {
  /**
       * The Docker image to use for this service.
       */
  readonly image: string;

  /**
      * The tag for the docker image.
      */
  readonly tag?: string;

  /**
       * Number of replicas to start.
       *
       * @default 1
       */
  readonly replicas?: number;

  /**
       * Secrets for deployment.
       *
       * @default undefined
       */
  readonly secret?: string;

  /**
      * Container commands.
      * 
      * @default []
      * 
      */
  readonly cmd?: string[];

  /**
       * External port.
       *
       * @default 80
       */
  readonly port?: number;

  /**
       * Extra env variables.
       *
       * @default []
       */
  readonly extraEnv?: { name: string, value: string }[];

  /**
       * Secret mounts for deployment container.
       *
       * @default []
       */
  readonly secretMounts?: { name: string, mountPath: string, subPath: string }[]

  /**
       * Internal port.
       *
       * @default port
       */
  readonly containerPort?: number;

  /**
      * Container pull policy. 
      * 
      * @default "IfNotPresent"
      * 
      */
  readonly pullPolicy?: 'IfNotPresent' | 'Always' | 'Never';

  /**
      * Liveliness Probe definitions for the container.
      */
  readonly livenessProbe?: probeProps

  /**
      * Readiness Probe definitions for the container.
      */
  readonly readinessProbe?: probeProps

}

export interface probeProps {
  /**
       * Sends a HTTP request to this path for the probe check. Only provide this OR a command.
       */
  readonly path?: string;

  /**
       * Comand to execute on the container for the probe. Only provide this OR a HTTP path.
       */
  readonly command?: string[];

  /**
       * Short for initialDelaySeconds: Number of seconds after the container has started before liveness or readiness probes are initiated. 
       * 
       * @default 0
       */
  readonly delay?: number;

  /**
       * Short for periodSeconds: How often (in seconds) to perform the probe. 
       * 
       * @default 10
       */
  readonly period?: number;
}


export class Container implements ContainerInterface {

  /**
       * Entrypoint array. 
       */
  readonly command?: string[];

  /**
       * List of environment variables to set in the container. 
       */
  readonly env?: EnvVar[];

  /**
       * List of sources to populate environment variables in the container. 
       */
  readonly envFrom?: EnvFromSource[];

  /**
       * Docker image name. 
       */
  readonly image?: string;

  /**
       * Image pull policy. 
       */
  readonly imagePullPolicy?: string;

  /**
       * Name of the container specified as a DNS_LABEL.
       */
  readonly name: string;

  /**
       * List of ports to expose from the container. 
       *
       */
  readonly ports?: ContainerPort[];

  /**
       * Pod volumes to mount into the container's filesystem.
       */
  readonly volumeMounts?: VolumeMount[];

  /**
       * Periodic probe of container service readiness.
       */
  readonly readinessProbe?: Probe;

  /**
       * Periodic probe of container liveness. 
       */
  readonly livenessProbe?: Probe;

  constructor(props: ContainerProps) {

    this.name = 'worker'
    // Priority is tag set, GIT_SHA env var, 'latest'
    const GIT_SHA = process.env.GIT_SHA || 'latest';
    const tag = props.tag || GIT_SHA;
    this.image = `${props.image}:${tag}`;
    this.ports = [{ containerPort: props.port ?? 80 }]
    this.imagePullPolicy = props.pullPolicy || 'IfNotPresent';
    this.command = props.cmd;
    this.volumeMounts = props.secretMounts;
    this.envFrom = props.secret ? [{ secretRef: { name: props.secret } }] : undefined;
    let envVars = [...(props.extraEnv || [])];
    envVars.push({ name: 'GIT_SHA', value: GIT_SHA });
    this.env = envVars;
    this.readinessProbe = props.readinessProbe && new Probe(props.readinessProbe);
    this.livenessProbe = props.livenessProbe && new Probe(props.livenessProbe);
  }

}

export class Probe implements ProbeInterface {

  /**
       * One and only one of the following should be specified. Exec specifies the action to take.
       */
  readonly exec?: ExecAction;

  /**
       * HTTPGet specifies the http request to perform.
       */
  readonly httpGet?: HttpGetAction;

  /**
       * Number of seconds after the container has started before liveness probes are initiated. 
       */
  readonly initialDelaySeconds?: number;

  /**
       * How often (in seconds) to perform the probe. Default to 10 seconds.
       *
       */
  readonly periodSeconds?: number;

  constructor(props: probeProps) {
    this.initialDelaySeconds = props.delay ?? 1;
    this.periodSeconds = props.period ?? 10;
    if (props.command) {
      this.exec = { command: props.command }
    } else if (props.path) {
      this.httpGet = { path: props.path, port: 80 }
    } else throw new Error('Must provide either probe command or HTTP path')
  }
}

export interface VolumeProps {

  /**
       * Secret volume name. 
       */
  readonly name: string;

  /**
       * Secret volume mountPath. 
       */
  readonly mountPath: string;

  /**
       * Secret volume subPath. 
       */
  readonly subPath: string;

}


export class Volume implements VolumeInterface {

  /**
       * Name of the volume
       */
  readonly name: string;

  /**
       * Secret, stored as a volume
       *
       */
  readonly secret?: SecretVolumeSource

  constructor(props: VolumeProps) {
    let mountString = (a: string) => a.toLowerCase().split('_').join('-');

    this.name = `${mountString(props.name)}-${mountString(props.subPath)}`
    this.secret = {
      secretName: props.name,
      items: [{
        key: props.subPath,
        path: props.subPath,
      }],
    }
  }


}

