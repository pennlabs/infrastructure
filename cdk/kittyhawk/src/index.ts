
export * from './application';
export * from './autoscaler';
export * from './certificate';
export * from './chart';
export * from './container';
export * from './cronjob';
export * from './deployment';
export * from './ingress';
export * from './service';
