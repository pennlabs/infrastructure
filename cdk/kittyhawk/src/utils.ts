export type NonEmptyArray<T> = [T, ...T[]];

export function nonEmptyMap<T, U>(
  arr: NonEmptyArray<T>,
  callbackfn: (value: T) => U
): NonEmptyArray<U> {
  return arr.map(callbackfn) as any as NonEmptyArray<U>;
}
