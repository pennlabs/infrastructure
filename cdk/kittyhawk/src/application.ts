import { Construct } from 'constructs';
import { Deployment, DeploymentProps } from './deployment';
import { Ingress, IngressProps } from './ingress';
import { Service, ServiceProps } from './service'
import { Certificate } from './certificate';

/**
 * Warning: Before editing any interfaces, make sure that none of the interfaces will have 
 * property names that conflict with each other. Typescript may not throw an error and it 
 * could cause problems.
 */
export interface ApplicationProps extends IngressProps, DeploymentProps,
  ServiceProps { }


export class Application extends Construct {
  constructor(scope: Construct, appname: string, props: ApplicationProps) {
    super(scope, appname);

    // We want to prepend the project name to the name of each component
    const release_name = process.env.RELEASE_NAME || 'undefined_release';
    const fullname = `${release_name}-${appname}`

    new Service(this, fullname, props);

    new Deployment(this, fullname, props);

    if (props.ingress) {
      new Ingress(this, fullname, props)

      new Certificate(this, fullname, props)
    }
  }
}

/**
 * Helper function that first checks to make sure that the environment variable array 
 * doesn't already contain the env var, then inserts it into the array.
 * @param envArray array of environment variables to insert into 
 * @param envKey name of the environment variable to insert
 * @param envValue value of the environment variable
 */
function insertIfNotPresent(envArray: { name: string, value: string }[], envKey: string, envValue: any) {
  const envSettingsModule = envArray?.filter(env => (env.name === envKey));
  if (envSettingsModule?.length > 0) {
    throw new Error(`${envKey} should not be redefined as an enviroment variable.`)
  }
  envArray.push({ name: envKey, value: envValue });

}
export interface DjangoApplicationProps extends ApplicationProps {
  /**
   * Domain of the application.
   */
  readonly domain: string;

  /**
   * Just the list of paths passed to the ingress since we already know the host.
   */
  readonly ingressPaths: string[];

  /**
   * DJANGO_SETTINGS_MODULE environment variable.
   */
  readonly djangoSettingsModule: string;
}

export class DjangoApplication extends Application {
  constructor(scope: Construct, appname: string, props: DjangoApplicationProps) {

    // Have to be careful here with references when mutating things
    let djangoExtraEnv = Array.from(props.extraEnv || []);

    // Insert DJANGO_SETTINGS_MODULE and DOMAIN
    insertIfNotPresent(djangoExtraEnv, 'DJANGO_SETTINGS_MODULE', props.djangoSettingsModule)
    insertIfNotPresent(djangoExtraEnv, 'DOMAIN', props.domain)

    // Configure the ingress using ingressPaths.
    const djangoIngress = [{ host: props.domain, paths: props.ingressPaths }]

    // If everything passes, construct the Application.
    super(scope, appname, {
      ...props,
      extraEnv: djangoExtraEnv,
      ingress: djangoIngress,
    });
  }
}

export interface ReactApplicationProps extends ApplicationProps {
  /**
   * Domain of the application.
   */
  readonly domain: string;

  /**
   * Just the list of paths passed to the ingress since we already know the host.
   */
  readonly ingressPaths: string[];

  /**
   * PORT environment variable for react. Default '80'.
   */
  readonly portEnv?: string;
}

export class ReactApplication extends Application {
  constructor(scope: Construct, appname: string, props: ReactApplicationProps) {

    // Have to be careful here with references when mutating things
    let reactExtraEnv = Array.from(props.extraEnv || []);

    // Insert DOMAIN and PORT as env vars.
    insertIfNotPresent(reactExtraEnv, 'DOMAIN', props.domain)
    insertIfNotPresent(reactExtraEnv, 'PORT', props.portEnv || '80')

    // Configure the ingress using ingressPaths.
    const reactIngress = [{ host: props.domain, paths: props.ingressPaths }]

    // If everything passes, construct the Application.
    super(scope, appname, {
      ...props,
      extraEnv: reactExtraEnv,
      ingress: reactIngress,
    });
  }
}

export class RedisApplication extends Application {
  constructor(scope: Construct, appname: string, redisProps: Partial<ApplicationProps>) {
    super(scope, appname, {
      ...redisProps,
      image: 'redis',
      tag: redisProps.tag || '6.0',
      port: redisProps.port || 6379,
    });
  }
}
