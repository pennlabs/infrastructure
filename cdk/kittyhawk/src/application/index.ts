import { Construct } from 'constructs';
import { Certificate } from '../certificate';
import { Deployment, DeploymentProps } from '../deployment';
import { Ingress, IngressProps } from '../ingress';
import { Service } from '../service';
import { ServiceAccount } from '../serviceaccount';
import { nonEmptyMap } from '../utils';

import DjangoApplication from './django';
import ReactApplication from './react';
import RedisApplication from './redis';

export { DjangoApplication };
export { ReactApplication };
export { RedisApplication };

/**
 * Warning: Before editing any interfaces, make sure that none of the interfaces will have
 * property names that conflict with each other. Typescript may not throw an error and it
 * could cause problems.
 */
export interface ApplicationProps {
  readonly ingress?: IngressProps;
  readonly deployment: DeploymentProps;
  readonly port?: number;
  /**
   * Creates a service account and attach it to any deployment pods.
   * serviceAccountName: release name
   */
  readonly createServiceAccount?: boolean;
}

export class Application extends Construct {
  constructor(scope: Construct, appname: string, props: ApplicationProps) {
    super(scope, appname);

    // We want to prepend the project name to the name of each component
    const release_name = process.env.RELEASE_NAME || 'undefined_release';
    const fullname = `${release_name}-${appname}`;

    new Service(this, fullname, props.port);

    if (props.createServiceAccount) {
      new ServiceAccount(this, `${appname}-${release_name}`, {
        serviceAccountName: release_name
      })
    }

    new Deployment(this, fullname, {
      ...props.deployment,
      ...(props.createServiceAccount ? { serviceAccountName: release_name } : {}),
    });

    if (props.ingress) {
      new Ingress(this, fullname, props.ingress);

      nonEmptyMap(props.ingress.rules, rule => {
        return new Certificate(this, fullname, rule);
      });
    }
  }
}