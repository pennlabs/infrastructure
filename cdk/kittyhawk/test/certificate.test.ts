
import { Construct } from 'constructs';
import { Certificate } from '../src';
import { failingTest} from './utils'



export function buildFailingCertificateChart(scope: Construct) {

  /** If domain is invalid certificate creation should throw an error. **/
  new Certificate(scope, 'serve', {
    ingress: [{host: 'pennlabsorg', paths: ['/']}],
  })
}

test('Certificate -- Failing', () => failingTest(buildFailingCertificateChart));

