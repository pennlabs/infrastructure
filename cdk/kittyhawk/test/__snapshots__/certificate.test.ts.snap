// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Certificate (Non Penn Labs Chart) -- Success 1`] = `
Array [
  Object {
    "apiVersion": "cert-manager.io/v1",
    "kind": "Certificate",
    "metadata": Object {
      "labels": Object {
        "app.kubernetes.io/component": "certificate",
        "app.kubernetes.io/name": "pennlabs-org",
      },
      "name": "pennlabs-org",
    },
    "spec": Object {
      "dnsNames": Array [
        "pennlabs.org",
        "*.pennlabs.org",
      ],
      "issuerRef": Object {
        "group": "cert-manager.io",
        "kind": "ClusterIssuer",
        "name": "wildcard-letsencrypt-prod",
      },
      "secretName": "pennlabs-org-tls",
    },
  },
]
`;

exports[`Certificate -- Success 1`] = `
Array [
  Object {
    "apiVersion": "cert-manager.io/v1",
    "kind": "Certificate",
    "metadata": Object {
      "labels": Object {
        "app.kubernetes.io/component": "certificate",
        "app.kubernetes.io/managed-by": "kittyhawk",
        "app.kubernetes.io/name": "pennlabs-org",
      },
      "name": "pennlabs-org",
    },
    "spec": Object {
      "dnsNames": Array [
        "pennlabs.org",
        "*.pennlabs.org",
      ],
      "issuerRef": Object {
        "group": "cert-manager.io",
        "kind": "ClusterIssuer",
        "name": "wildcard-letsencrypt-prod",
      },
      "secretName": "pennlabs-org-tls",
    },
  },
]
`;
